### ISSUE TITLE
Occasion poor video quality 

### ISSUE THREAD

**Issue by theDude30** [twilio/twilio-video.js#119]
- [X ] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [ X] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [ X] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.

**Code to reproduce the issue:**

* Audio and video two ways.
* PC chrome to PC chrome.
* Unstable internet connection.

**Expected behavior:**

In case of unstable or poor internet connection i would expect some audio latency or some slower frame rate, maybe reduce video quality but should continue showing a visible video.

**Actual behavior:**

The video images were messed up(see attached):
![bad_video_example](https://cloud.githubusercontent.com/assets/5785389/26527576/33b132ae-439f-11e7-90a3-7b149cc06700.png)

Below are logs(with debug:true in the connect):
First PC:
---------------
console log:
[Uploading log_expert.txt…]()

webrtc-internals:
[webrtc_internals_dump.txt](https://github.com/twilio/twilio-video.js/files/1034047/webrtc_internals_dump.txt)

Second PC:
-----------------
Console log:
[customer.txt](https://github.com/twilio/twilio-video.js/files/1034049/customer.txt)

webrtc-internals:
[customer_webrtc_internals_dump.txt](https://github.com/twilio/twilio-video.js/files/1034051/customer_webrtc_internals_dump.txt)

Questions:
----------
* Can i restrict resolution/frame rate?
* Can i restrict the bandwidth?
* Will any of the restrictions could help?

**Software versions:**

 - [X ] Browser(s): Chrome 58.0.3029.110 (64-bit)
 - [X ] Operating System: OSX 10.12.5 (16F73)
 - [X] twilio-video.js: 1.0.0
 - [ ] Third-party libraries (e.g., Angular, React, etc.):
None

**Comment by markandrus** [comment]
Thanks, @theDude30. It looks like you're using Group Rooms here, i.e. your video is flowing through Twilio's Media Server/SFU rather than P2P. The corrupted frames you see _might_ be a known bug in the Media Server. I am raising it to the team. As to your questions...

> Can i restrict resolution/frame rate?

Yes, but you must do this when creating your LocalVideoTrack, i.e. you can't really do it once connected to the Room, unless you replace the Track (i.e., remove existing Track, add new Track). Here are the constraints you can pass:

```js
createLocalVideoTrack({
  width: 320,
  height: 240,
  frameRate: 15
});
```

In the future, once we support simulcast/scalable video coding (SVC), these tradeoffs could be performed more or less automatically.

> Can i restrict the bandwidth?

We are working on an addition to ConnectOptions so that you could constrain the bandwidth on the underlying PeerConnection(s). It might look something like this:

```js
connect(token, {
  bandwidth: 500 // 500 Kbps
});
```

Although I'm not sure what units we should specify in. Do you have any feedback/requests for such an API?

> Will any of the restrictions could help?

Yes, they could help in some cases; however, as I mentioned above, I think you may have hit a known bug in the media server. I'm waiting on confirmation from that team.

Thanks,
Mark

**Comment by theDude30** [comment]
I need to make 1 on 1 audio/video in a specific room.
Is there a way to connect/create a room and use P2P or relayed audio/video without any media server involved ?
it seems that 5 out of 20 videos are getting corrupted in a way that it is not usable and all of those environments have great network bandwidth.

**Comment by markandrus** [comment]
@theDude30 one way you could do this is by setting the Default Room Settings in your account—e.g., you could select "Peer-to-peer" as your default Room Type:

<img alt="screen shot 2017-06-01 at 9 08 19 am" src="https://cloud.githubusercontent.com/assets/359154/26689243/e13e6a4c-46a9-11e7-9159-6b2eb97339d9.png">

Do you need to mix both Peer-to-peer and Group Rooms?

**Comment by theDude30** [comment]
Cool. 
I wasn't aware that i can choose between topology, is this a new feature?
I have small amount of participants per room(1 to 3) so i think "peer-to-peer" only should do the job.
 
Thanks.

**Comment by markandrus** [comment]
@theDude30 yes, the new Group Room Type was released May 18th and coincided with a [blog post](https://www.twilio.com/blog/2017/05/announcing-programmable-video-group-rooms.html). When we rolled out the Group Room Type, we updated the Default Room Settings to allow you to select between Peer-to-peer and Group Rooms in the Console, defaulting to Group Rooms for new accounts. If you had an account before May 18th, the default there should have remained Peer-to-peer. If you opened an account on or after May 18th, the default should have been set to Group Rooms.

Regarding the Video corruption, our Media Server Team is looking into it. They have a couple leads, one of which may point to an issue in Chrome. They are still researching.

**Comment by theDude30** [comment]
I switched the account for "peer to peer" and we still see video corruption from time to time.

**Comment by markandrus** [comment]
@theDude30 this might point to the potential issue in Chrome I mentioned. Our media team is still looking into it, but thanks for this contribution--it narrows things down.

**Comment by theDude30** [comment]
do you have any estimates on when is it going to be fixed?
We are using this product on production and it makes it unusable in 25% of all audio/video calls.

**Comment by mparis** [comment]
Hello @theDude30, this is a known issue in Chrome M58 fixed by [1].
It should have been fixed in Chrome M59, but we are already dealing with corrupted video problems in SFU topology as reported in [2], but we do not know yet if it happens in P2P.
If you have the chance, I suggest to use Chrome M59 (right now is in beta) in order to:

- Reduce the frequency of the bug (after some tests it seems that it has been reduced)
- If you have problems in P2P topology, could you please share it with us in order to have more info to definitely fix the bug?

Thanks!!

Refs
[1] https://chromium.googlesource.com/external/webrtc/trunk/webrtc.git/+/9d4241cbf21ab48686b80918bf5e4d72f2c15574
[2] https://bugs.chromium.org/p/webrtc/issues/detail?id=7768

**Comment by mparis** [comment]
Hello @theDude30,
are you using P2P Rooms?

**Comment by theDude30** [comment]
Yes, the below screenshot is the room setting that we are using:
<img width="1131" alt="screen shot 2017-06-28 at 20 38 44" src="https://user-images.githubusercontent.com/5785389/27651710-ef82bd80-5c41-11e7-9f49-e7157889c4be.png">

**Comment by theDude30** [comment]
Correction: the room type is GROUP not peer-to-peer as mentioned above.

**Comment by mparis** [comment]
OK, thanks for the feedback ;).
